sdk/core/azure-core/azure/core/messaging.py

```diff
@@ -0,0 +1,132 @@
+    :type type: str
+    :keyword data: Optional. Event data specific to the event type. If data is of bytes type, it will be sent
+     as data_base64 in the outgoing request.
+    :type data: object
+    :keyword time: Optional. The time (in UTC) the event was generated, in RFC3339 format.
+    :type time: ~datetime.datetime
+    :keyword dataschema: Optional. Identifies the schema that data adheres to.
+    :type dataschema: str
+    :keyword datacontenttype: Optional. Content type of data value.
+    :type datacontenttype: str
+    :keyword subject: Optional. This describes the subject of the event in the context of the event producer
+     (identified by source).
+    :type subject: str
+    :keyword specversion: Optional. The version of the CloudEvent spec. Defaults to "1.0"
+    :type specversion: str
+    :keyword id: Optional. An identifier for the event. The combination of id and source must be
+     unique for each distinct event. If not provided, a random UUID will be generated and used.
+    :type id: Optional[str]
+    :ivar source: Identifies the context in which an event happened. The combination of id and source must
+     be unique for each distinct event. If publishing to a domain topic, source must be the domain name.
+    :vartype source: str
+    :ivar data: Event data specific to the event type.
+    :vartype data: object
+    :ivar type: Type of event related to the originating occurrence.
+    :vartype type: str
+    :ivar time: The time (in UTC) the event was generated, in RFC3339 format.
+    :vartype time: ~datetime.datetime
+    :ivar dataschema: Identifies the schema that data adheres to.
+    :vartype dataschema: str
+    :ivar datacontenttype: Content type of data value.
+    :vartype datacontenttype: str
+    :ivar subject: This describes the subject of the event in the context of the event producer
+     (identified by source).
+    :vartype subject: str
+    :ivar specversion: Optional. The version of the CloudEvent spec. Defaults to "1.0"
+    :vartype specversion: str
+    :ivar id: An identifier for the event. The combination of id and source must be
+     unique for each distinct event. If not provided, a random UUID will be generated and used.
+    :vartype id: Optional[str]
+    """
+    def __init__(self, source, type, **kwargs): # pylint: disable=redefined-builtin
+        # type: (str, str, Any) -> None
+        self.source = source
+        self.type = type
+        self.specversion = kwargs.pop("specversion", "1.0")
+        self.id = kwargs.pop("id", str(uuid.uuid4()))
+        self.time = kwargs.pop("time", datetime.now(TZ_UTC).isoformat())
+        self.datacontenttype = kwargs.pop("datacontenttype", None)
+        self.dataschema = kwargs.pop("dataschema", None)
+        self.subject = kwargs.pop("subject", None)
+        self.extensions = {}
+        _extensions = dict(kwargs.pop('extensions', {}))
+        for key in _extensions.keys():
```

Comment:
Does this validation show up on a profiler if you create/send a lot of messages?

Reply:
While this check matches the spec indeed, I wonder if we should do client side validation here?

Reply:
Interesting question - ideally the service must validate it (it doesn't right now)

Reply:
So does the data get dropped? Something else?

Reply:
right now there are two options i see- 
- either we validate on the client side to adhere to the spec
- we don't validate and send it to the service(s) as-is and let them take care of it. 

I incline to the fist approach, but i'm okay with the second (for perf reasons)

Reply:
if we are going with the second, I'm open to skip validation by just documenting the spec of the extension attributes

Reply:
Will the service reject the call or will it silently drop data that has invalid keys (names)?

Reply:
the service (eventgrid) will accept any extensions - but we are most likely going to have services that use cloud events and might not accept invalid extension keys

We have decided to add these validations cross language for 
1) Perf isn't really that big of a concern here
2) If someone really complains, we can always play the "additive is not breaking" card and add those later

Reply:
Changing validation (and thus what exception gets raised) is not an "additive is not breaking" type of change. That would be a behavioral change. You may be able to slot it under "incorrect data in, slightly different failure out". But, for example, if you start raising exceptions (possibly from the service) that are not `ValueError`s, then a previously working client can very much go boom.

Reply:
all the more reason to stick to the cloudevent spec? we must assume multiple services WILL use this from azure core - and each service will have their own behavior - and it's not our responsibility to dictate what service does - so sticking to the spec is the safest bet we can have?

alternatively, if the users really don't want validation on anything, they are free to use dictionaries

Reply:
I'm good with that rationale.